dali/operators/signal/decibel/to_decibels_op_gpu.cu

```diff
@@ -0,0 +1,138 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <utility>
+#include <vector>
+#include <memory>
+#include "dali/operators/signal/decibel/to_decibels_op.h"
+#include "dali/core/static_switch.h"
+#include "dali/kernels/kernel_params.h"
+#include "dali/kernels/reduce/reduce_all_kernel_gpu.h"
+#include "dali/kernels/signal/decibel/to_decibels_gpu.h"
+#include "dali/pipeline/data/views.h"
+
+namespace dali {
+
+template <typename T, int Dims>
+class ToDecibelsImpl : public OpImplBase<GPUBackend> {
+ public:
+  using MaxKernel = kernels::reduce::ReduceAllGPU<T, T, Dims, kernels::reductions::max>;
+  using ToDecibelsKernel = kernels::signal::ToDecibelsGpu<T, Dims>;
+  using ToDecibelsArgs = kernels::signal::ToDecibelsArgs<T>;
+
+  explicit ToDecibelsImpl(ToDecibelsArgs args)
+      : args_(std::move(args)) {
+    if (args_.ref_max) {
+      kmgr_max_.Initialize<MaxKernel>();
+      kmgr_max_.Resize<MaxKernel>(1, 1);
+    }
+    kmgr_todb_.Initialize<ToDecibelsKernel>();
+    kmgr_todb_.Resize<ToDecibelsKernel>(1, 1);
+  }
+
+  bool SetupImpl(std::vector<OutputDesc> &output_desc, const workspace_t<GPUBackend> &ws) override;
+  void RunImpl(workspace_t<GPUBackend> &ws) override;
+
+ private:
+  ToDecibelsArgs args_;
+
+  kernels::KernelManager kmgr_max_;
+  kernels::KernelManager kmgr_todb_;
+
+  std::vector<OutputDesc> max_out_desc_;
+  TensorList<GPUBackend> max_out_;
+};
+
+template <typename T, int Dims>
+bool ToDecibelsImpl<T, Dims>::SetupImpl(std::vector<OutputDesc> &output_desc,
+                                        const workspace_t<GPUBackend> &ws) {
+  kernels::KernelContext ctx;
+  ctx.gpu.stream = ws.stream();
+
+  const auto &input = ws.InputRef<GPUBackend>(0);
+  auto in_view = view<const T, Dims>(input);
+  auto nsamples = input.size();
+
+  auto type = TypeInfo::Create<T>();
+
+  if (args_.ref_max) {
+    auto& req_max = kmgr_max_.Setup<MaxKernel>(0, ctx, in_view);
+    max_out_desc_.resize(1);
+    max_out_desc_[0].type = type;
+    max_out_desc_[0].shape = req_max.output_shapes[0];
+  }
+
+  auto& req = kmgr_todb_.Setup<ToDecibelsKernel>(0, ctx, in_view);
+  output_desc.resize(1);
+  output_desc[0].type = type;
+  output_desc[0].shape = req.output_shapes[0];
+
+  return true;
+}
+
+template <typename T, int Dims>
+void ToDecibelsImpl<T, Dims>::RunImpl(workspace_t<GPUBackend> &ws) {
+  const auto &input = ws.InputRef<GPUBackend>(0);
+  auto &output = ws.OutputRef<GPUBackend>(0);
+  int nsamples = input.size();
+
+  auto in_view = view<const T, Dims>(input);
+  auto out_view = view<T, Dims>(output);
+
+  kernels::KernelContext ctx;
```

Comment:
Can't we share the ctx between setup and run? I see that we set stream twice for ctx, it looks error prone.

Reply:
Actually, we don't need the stream during Setup, so I'll remove it.
Regarding sharing context, I think @mzient's intention was KernelContext to be a small object that you create every time, rather than keeping it in the instance. Am I right Michal?

Reply:
@mzient - so what is the purpose of the context if we are keeping recreating it and sometimes it doesn't carry any state.

Reply:
it's the context the kernel needs to run. We are also passing it to the Setup function but in most cases if not all it is not needed there

Reply:
> We are also passing it to the Setup function

My reservation is that the ctx is different between setup and run, and I think the idea was to share a state between these two functions.

Reply:
Actually, it's mostly for CUDA stream. If we decide to go for some suballocators (and not making them global), they would be passed this way, too. It's not for Kernel state.